### PR TITLE
perf: reduce SSE flush delay and add buffered writer

### PR DESCRIPTION
**Pull request by srstack**
## Summary

This PR optimizes SSE event delivery latency by introducing a buffered writer and reducing the flush delay.

## Changes

- **Add `sseBufferedWriter`**: Wraps `http.ResponseWriter` with `bufio.Writer` to batch small SSE writes
- **Reduce flush delay**: From 5ms to **1ms** (`sseFlushMaxDelay`)
- **Batch size**: Flush after 10 events (`sseFlushBatchSize`) or 1ms max delay
- **Heartbeat flush**: Maintains existing heartbeat-based flush for reliability

## Performance Impact

Based on production testing against `api.drive9.ai`:

| Metric | Before | After (expected) |
|--------|--------|-----------------|
| SSE Connect | 103ms | 103ms |
| SSE Latency (P50) | ~790ms | ~100-200ms |
| Reconnect Loss | 100% | 0% (fixed via since param) |

## Testing

- All existing `pkg/server` tests pass
- Production smoke test: 30 events, 29/30 received, avg latency 791ms (will improve after deployment)

## Backward Compatibility

- No API changes
- Client-side `since` parameter already supported for replay
- EventBus ring buffer (10k events) unchanged

## Summary by CodeRabbit

* **Refactor**
  * Improved real-time event delivery by introducing buffered SSE writes that batch events and coalesce bursts, with heartbeats and timed flush boundaries to reduce overhead and improve throughput.

* **Bug Fixes**
  * Replay/reset notifications (stale-cursor) are flushed immediately to ensure correct client sync; flush behavior is robust against timer races.

* **Tests**
  * Added tests for batch-size flush, max-delay flush, explicit flush, burst flush timing, and reset-on-stale-cursor.

### CODE REVIEW

**Comment by srstack** [comment]
Fixed both review comments:

1. **Duplicate counters** (`pending` vs `bw.count`): Removed the `pending` variable in `handleEvents`. Now uses `bw.count` as the single source of truth. `bw.shouldFlush()` handles both batch-size and max-delay checks consistently.

2. **Unchecked type assertion**: Changed `newSSEBufferedWriter` to explicitly accept `http.Flusher` as a parameter. The caller (`handleEvents`) already checks `w.(http.Flusher)` and passes the verified flusher directly, eliminating the panic risk.

Please re-review.

**Comment by srstack** [comment]
Fixed all new review comments:

1. **Timer initialization race**: Replaced `time.NewTimer(0).Stop()` with lazy timer allocation. Added `stopTimer()` helper with safe stop+drain pattern.

2. **Flush error handling**: `doFlush` now checks and handles `bw.Flush()` errors. Handler returns early on broken pipe instead of looping silently.

3. **Reset not flushed immediately**: When `liveOK` is false, reset is now flushed immediately via `bw.Flush()` instead of being buffered until next heartbeat.

4. **Missing test for seq_too_old**: Added `TestSSEResetFlushWhenSeqTooOld` to verify reset is flushed promptly when ring buffer wraps.

5. **Timer Stop without drain**: All `flushTimer.Stop()` calls now use the `stopTimer()` helper that drains the channel.

6. **Max delay test assertion**: `TestSSEBufferedWriterMaxDelayFlush` now asserts `shouldFlush()` becomes true after the delay.

Please re-review.

**Comment by srstack** [comment]
Fixed all new review comments:

1. **recordWrite on failed writes**: Now only calls `recordWrite()` when `fmt.Fprintf` succeeds (err == nil).

2. **Timer Reset without stop+drain**: Added `stopTimer(flushTimer)` before `Reset()` calls.

3. **Heartbeat not stopping timer**: Heartbeat path now stops timer and sets `flushC = nil` to avoid spurious wakeups.

4. **Inline function in select**: Replaced with `var flushC <-chan time.Time` that is updated when timer is created/stopped.

5. **Comment mismatch**: Updated type comment to remove "heartbeat resets the timer" and test comment to match actual since=4 value.

Please re-review.

**Comment by srstack** [comment]
Fixed CodeRabbit review: coalesced burst timer arming bug.

**Bug:** When multiple `Publish` calls coalesce into one `notify` wakeup, `EventsSince()` returns N>1 events. The old code only armed the timer when `bw.count == 1`, so bursts of 2-9 events would sit buffered until the next heartbeat (30s) instead of flushing within 1ms.

**Fix:** Changed timer arming condition from `bw.count == 1` to `flushC == nil`. Now any non-empty buffer (whether from 1 event or a burst of N) arms the max-delay timer.

**Test:** Added `TestSSEBurstFlush` which publishes 3 events concurrently and asserts the client receives them within 200ms (well under heartbeat interval).

All 12 SSE tests pass.

Please re-review.

File: pkg/server/sse.go [hunk 1]
```diff
@@ -112,42 +176,60 @@ func (s *Server) handleEvents(w http.ResponseWriter, r *http.Request) {
...
 				}
 			}
-			flusher.Flush()
+			if bw.count > 0 {
+				if bw.shouldFlush() {
+					doFlush()
+					flushTimer.Stop()
+				} else if bw.count == 1 {
+					// First buffered event starts the max-delay timer.
+					flushTimer.Reset(sseFlushMaxDelay)
```
Comment (qiffang) [inline pkg/server/sse.go]: 🔴 **Blocking bug**：这里只在 `bw.count == 1` 时启动 timer。但 `bus.EventsSince(lastSeen)` 一次可能返回 N 条事件（live notify 经常一次推 batch）。当 1 < N < 10 时：

- `shouldFlush()` 返回 false（< batchSize）
- `bw.count == 1` 也是 false（== N）
- **timer 完全不会启动**，这 N 条事件就卡在 buffer 里直到下一次 notify 凑够 10 条，或下一个 30s heartbeat 顺手 flush

实际效果：把 "每事件 ≤790ms" 退化成 "小批量 ≤30s"，比改前更糟。

修法（保留 timerArmed 状态，只要有 buffered 数据且 timer 没在跑就 arm）：
```go
if bw.count > 0 {
    if bw.shouldFlush() {
        doFlush()
        flushTimer.Stop()
        timerArmed = false
    } else if !timerArmed {
        flushTimer.Reset(sseFlushMaxDelay)
        timerArmed = true
    }
}
```
注意 Reset 之前要 drain（见 line 195 评论）。

配套补一个回归测试：连续 publish 5 条 event，验证客户端在 << 30s 内拿到全部 5 条。

File: pkg/server/sse.go [hunk 2]
```diff
@@ -112,42 +176,60 @@ func (s *Server) handleEvents(w http.ResponseWriter, r *http.Request) {
...
 
-	// Phase 2: Live stream.
+	// Phase 2: Live stream with micro-batching.
+	// Instead of flushing after every single event, we accumulate events
+	// and flush at heartbeat boundaries or when the batch size is reached.
 	heartbeat := time.NewTicker(sseHeartbeatInterval)
 	defer heartbeat.Stop()
 
+	flushTimer := time.NewTimer(0)
+	flushTimer.Stop()
```
Comment (qiffang) [inline pkg/server/sse.go]: 🔴 **Timer drain race**：`time.NewTimer(0)` 立刻 fire，channel 里已写入信号；接着 `flushTimer.Stop()` 返回 false 但**不 drain**。

后续在 select 循环里如果先进了 `case <-flushTimer.C` 没事（`bw.count==0`空跑），但下一次 `flushTimer.Reset(...)` 在未 drain 的 channel 上行为依赖 Go 版本（Go 1.23 之前会 race）。

建议改成：
```go
flushTimer := time.NewTimer(0)
if !flushTimer.Stop() {
    <-flushTimer.C
}
```
或干脆别用 0 长 timer，改成 `var flushTimer *time.Timer` + 第一次需要时 `flushTimer = time.NewTimer(d)`，省掉 drain。

File: pkg/server/sse.go [hunk 3]
```diff
@@ -14,8 +15,65 @@ import (
...
+	w       *bufio.Writer
+	count   int
+	first   time.Time
+	flusher http.Flusher
+}
+
+func newSSEBufferedWriter(rw http.ResponseWriter, flusher http.Flusher) *sseBufferedWriter {
+	return &sseBufferedWriter{
+		rw:      rw,
+		w:       bufio.NewWriter(rw),
```
Comment (qiffang) [inline pkg/server/sse.go]: 🟠 **bufio.Writer 自动 flush 不调 http.Flusher**：`bufio.NewWriter` 默认 buffer 4096 字节。当 buffered 数据超过 4 KiB（10 条 event 如果 path 较长很容易超），bufio **自己**会 write 到 underlying ResponseWriter，**但不会**调用 `flusher.Flush()`。

这意味着虽然 bufio 已经把数据丢给 chunked encoder，HTTP 层依然可能停在 net/http 的写 buffer / Nginx / TLS 缓冲里 —— 客户端看不到。

两条修法：
1. `bufio.NewWriterSize(rw, 8192)` 让 buffer 比一个 batch 大，加上 explicit `Flush()` 兜底；
2. 自定义一层 wrapper：每次 bufio.Writer 把数据吐到 underlying writer 时同步调 `http.Flusher.Flush()`（更稳妥）。

建议方案 2 + 加一个测试：写一条很大的 event（> 4 KiB payload）验证 client 立刻能收到，不依赖外层 explicit Flush。

File: pkg/server/sse.go [hunk 4]
```diff
@@ -14,8 +15,65 @@ import (
 
 const (
 	sseHeartbeatInterval = 30 * time.Second
+	sseFlushBatchSize    = 10
+	sseFlushMaxDelay     = 1 * time.Millisecond
```
Comment (qiffang) [inline pkg/server/sse.go:19]: 🟡 **常量值合理性**：`sseFlushMaxDelay = 1ms` 极激进，几乎等于不批量。`sseFlushBatchSize = 10` 在 1ms 内基本攒不到。

这两个值组合下，绝大多数 batch 由 timer 而非 batchSize 触发，本 PR 的"批量化"优化效果会大打折扣，更多是"每事件 +1ms 延迟"而已。

建议：
- 要么把 maxDelay 提到 5-20ms（结合 user-facing latency 可接受范围）；
- 要么增加一个测试或 production metric 证明当前值在生产 event rate 下确实 hit batchSize 路径。

额外：定数值没注释来源（为啥是 10 / 1ms 而不是 50 / 5ms？），建议 godoc 一两行说明。

File: pkg/server/sse_test.go [hunk 5]
```diff
@@ -316,3 +316,93 @@ func TestSSEEndpointBadSince(t *testing.T) {
...
+
+	// After batch flush, all 10 events should be in the body.
+	body := rec.Body.String()
+	count := strings.Count(body, "event: heartbeat")
+	if count != 10 {
+		t.Fatalf("expected 10 heartbeat events after batch flush, got %d", count)
+	}
+}
+
+func TestSSEBufferedWriterMaxDelayFlush(t *testing.T) {
```
Comment (qiffang) [inline pkg/server/sse_test.go:361]: 🟡 **这个测试没真覆盖 timer 路径**：注释明写 "In production the timer calls Flush; here we simulate it." —— 也就是只测了直接调 `Flush()` 同步函数，**没**触发生产代码里 `case <-flushTimer.C:` 那条分支。

而上面提到的"notify 多事件 timer 不启动"这个 bug 完全在 `handleEvents` 控制流里，单元测 `sseBufferedWriter` 测不到。

建议改成集成测：起一个 `httptest.Server` 跑 `handleEvents` → 通过 `bus.Publish` 一次性推 5 条 event → 在客户端用短超时（比如 50ms）验证 5 条都到了。这才能覆盖关键 bug。

**Review by qiffang — CHANGES_REQUESTED** [review]
## Review Summary

这个 PR 在思路上是对的（buffer 多个 SSE event 减少 syscall + flush），但**当前实现有一个会让 user-facing latency 反而比改之前更糟的 bug**，再加上几个误导性的描述/测试缺口，建议先修后再合。

### 🔴 Blocking — Phase 2 notify 分支的延迟刷新逻辑

`pkg/server/sse.go:224-231` 的判断只在 `bw.count == 1` 时启动 `flushTimer`。但 `EventsSince(lastSeen)` 一次可以返回多条事件（典型场景：客户端刚连上、live notify 唤醒时本来就经常一次推 N 条）。

如果一次 notify 写入了 2 ≤ N < 10 条 event：
- `shouldFlush()` 返回 false（count < 10）
- `bw.count == 1` 也是 false（count == N > 1）
- **timer 不会被启动，事件永远卡在 buffer 里**，直到下一次 notify 凑够 10 条，或下一个 30 秒 heartbeat 顺手 flush

相当于把"每个 event ≤790ms 抵达"退化成"小批量 event 最坏 30s 抵达"。这正好是 PR 想要解决的反面。

建议改成：
```go
if bw.count > 0 {
    if bw.shouldFlush() {
        doFlush()
        flushTimer.Stop()
        timerArmed = false
    } else if !timerArmed {
        // 任何首条 buffered event（不只是 count==1）都启动 timer
        flushTimer.Reset(sseFlushMaxDelay)
        timerArmed = true
    }
}
```

### 🔴 Blocking — `flushTimer` 的 Stop / Reset 没有 drain

`pkg/server/sse.go:195-197`：
```go
flushTimer := time.NewTimer(0)
flushTimer.Stop()
```
`NewTimer(0)` 立刻 fire，channel 里已经写入信号；`Stop()` 返回 false 但 **不 drain channel**。然后 select 循环会立刻拿到一次 spurious `flushTimer.C`，因为 `bw.count == 0` 走空分支看起来 OK，但后续 `flushTimer.Reset(...)` 在 channel 未消费的情况下行为依赖 Go 版本（Go ≥ 1.23 修了，但 < 1.23 就有 race）。

建议显式 drain：
```go
flushTimer := time.NewTimer(0)
if !flushTimer.Stop() {
    <-flushTimer.C
}
```
或直接 `var flushTimer *time.Timer` + 第一次需要时 `flushTimer = time.NewTimer(d)`，省掉 drain。

### 🟠 Major — `bufio.Writer` 自动 flush 不会触发 `http.Flusher.Flush()`

`bufio.Writer` 默认 4096 buffer。当 buffered 数据超过 4 KiB（10 条事件如果 path 较长就够了）时，bufio **自己**会 write 到 underlying `http.ResponseWriter`，但**不会**调用 `flusher.Flush()`。这意味着即便 bufio 已经把数据交给 chunked encoder，**HTTP 层在没显式 Flush 之前数据仍可能停在 net/http 内部 buffer / Nginx / TLS 写缓冲里**，客户端依旧看不到。

解决：
- `bufio.NewWriterSize(rw, 8192)` 让 buffer 比一个 batch 大，加上 explicit `Flush()` 兜底；
- 或自定义一层 wrapper：每次 `bufio` flush 到 underlying writer 时同步调 `http.Flusher.Flush()`（更稳妥）。

### 🟠 Major — PR 描述与 diff 不符

PR description 表格写：
```
| Reconnect Loss | 100% | 0% (fixed via since param) |
```
但 diff 完全没改 `since` 相关逻辑（client/server 都没动）。这条要么删掉，要么独立另一个 PR；现在挂在这里有误导。

另外 "SSE Latency P50 ~790ms → ~100-200ms" 也很奇怪：当前 max delay 是 1ms，理论上 P50 不应该被这个 1ms 主导到 100-200ms 量级。如果 production 量出来真是 100-200ms，**说明 latency 主因不在 server flush 频率**，而在别的链路（DNS/TLS/Nginx/proxy/客户端 bufio）。这点请在 description 里厘清。

### 🟡 Nitpick — 测试覆盖不到核心 bug

`TestSSEBufferedWriterMaxDelayFlush` 注释明确写着 "In production the timer calls Flush; here we simulate it." —— 也就是说**完全没真测到 timer 路径**，只测了 `Flush()` 同步函数。上面 🔴 #1 那个 "小批量被卡 30s" 的 bug 一行测试都没覆盖。

建议补：
- `TestHandleEventsLiveBatchUnderTen`: 启 handleEvents goroutine，连续 publish 5 个 event，验证 client 在 ≤ 5×sseFlushMaxDelay 内拿到全部 5 条（race detector 下）。
- `TestHandleEventsBufferedFlushOnHeartbeat`: 验证 heartbeat 来时会刷 buffered events。
- `TestSSEBufferedWriterLargePayload`: 写 > 4 KiB 验证 bufio 自动溢出时 client 仍能拿到（暴露 🟠 #1）。

### 🟡 Nitpick — 这个 PR 不解决 reset storm

实际线上观察到 `dat9: SSE reset — invalidating all caches` 在 actor 自己的 100 个 PUT 完成后被刷屏 102 次（juicefs bench 触发）。原因是 `isStructuralOp` 把任何 rename/delete/mkdir/copy 都直接转成 reset，而 client 的 `handleEvent` 不对 reset 做 self-actor filter（pkg/fuse/sse.go），自家 mount 制造的 namespace churn 也会自食。

本 PR 是 "把 reset 也批量化"，**没有解决 reset 频次问题本身**。请在 description 里明确：本 PR 只降 flush 延迟，不解决 structural-change reset storm。reset storm 需要另开 PR（候选方向：client 对 reset 也做 actor filter，或 server 把 self-actor 的 structural change 用 file_changed + 强 invalidate 路径替代 reset）。

### 选项汇总

- **必须修：** 🔴 #1（notify 多事件 timer 不启动）、🔴 #2（timer drain）
- **建议修：** 🟠 #1（bufio 自动 flush 不触发 http.Flusher）、🟠 #2（PR 描述对齐）
- **建议补：** 🟡 #1（关键路径回归测试）、🟡 #2（在 description 标注本 PR 不解决 reset storm）

修了 🔴 之后我倾向 approve；建议补 🟡 #1 一起进，避免下次回归。

**Comment by srstack** [comment]
@qiffang Thanks for the detailed review. All blocking issues have been addressed:

**🔴 Blocking #1 (coalesced burst timer)**: Fixed. Changed timer arming condition from `bw.count == 1` to `flushC == nil` so any non-empty buffer (whether 1 event or a burst of N) arms the max-delay timer. Added `TestSSEBurstFlush` which publishes 3 events concurrently and asserts the client receives them within 200ms.

**🔴 Blocking #2 (timer drain)**: Fixed. Uses lazy `*time.Timer` allocation + `stopTimer()` helper with safe stop+drain pattern.

**🟠 Major #1 (bufio auto flush)**: Fixed. Added `flusherWriter` wrapper that calls `http.Flusher.Flush()` on every underlying write. Also increased bufio buffer to 64 KiB to prevent auto-flush within a batch.

**🟠 Major #2 (PR description)**: Updated. Removed the misleading "Reconnect Loss" claim. Added "Not in this PR" section clarifying that this PR only reduces flush latency, not reset storm frequency.

**🟡 Nitpick #1 (test coverage)**: Added `TestSSEBurstFlush` (coalesced burst timer path) and `TestSSEResetFlushWhenSeqTooOld` (immediate reset flush path). The existing `TestSSEEndpointLiveEvent` already covers the basic live event path.

Please re-review.

**Comment by qiffang** [comment]
Reviewed this PR in detail across the server path (`pkg/server/sse.go`) and the current consumers (`pkg/client/events.go`, `pkg/fuse/sse.go`, and the Obsidian watcher).

I do not see a correctness regression here.

Why this looks safe:
- The wire format and event ordering are unchanged. This PR changes when bytes are flushed, not the SSE event schema or seq progression.
- Initial replay/reset is still flushed immediately, so a fresh/reconnecting client still gets the correct cursor position right away.
- The `seq_too_old` live reset path is now explicitly flushed immediately, which avoids stalling a reconnecting client behind the 30s heartbeat.
- Current clients parse by SSE event boundaries (`event: ...` / `data: ...` / blank line), not by per-write flush boundaries, so coalescing multiple events into one network flush should not break them.
- The added tests cover the important edge cases introduced by batching: batch threshold flush, max-delay flush, burst delivery, and wrapped-ring `seq_too_old` reset.

Net: no blocking issues from my side.

**Review by qiffang — APPROVED** [review]
